### DISCUSSION TITLE
Cohesive Zone Model : Obtain AuxVariable Neighbor Information

### DISCUSSION THREAD

**Discussion by chunhuizhao478** [idaholab/moose#22599] [Q&A Modules: Solid mechanics] [answered]
Hi all, we try to implement custom CZM material kernel which needs reactions as variable and calculate the difference of each interface node pair from the plus and minus side. After digging into the discussion, I realize there is an easy way to save the reactions as auxvariable by using "save_in_master" and "save_in_slave" , the input file looks something like:

````
[AuxVariables]
   [./resid_primary_x]
      order = FIRST
      family = LAGRANGE
    [../]
    [./resid_primary_y]
      order = FIRST
      family = LAGRANGE
    [../]
    [./resid_secondary_x]
      order = FIRST
      family = LAGRANGE
    [../]
    [./resid_secondary_y]
      order = FIRST
      family = LAGRANGE
    [../]
[]

[Modules/TensorMechanics/CohesiveZoneMaster]
    [./czm_ik]
      boundary = 'Block0_Block1'
      strain = SMALL
      generate_output='traction_x traction_y jump_x jump_y normal_traction tangent_traction normal_jump tangent_jump'
      save_in_master = 'resid_primary_x resid_primary_y'
      save_in_slave  = 'resid_secondary_x  resid_secondary_y' 
    [../]
 []
````

Then I feed residuals into the material kernel and inside `computeInterfaceTractionAndDerivatives`, I would like to compute something like:

````
resid_primary_x[_qp] - residuals_secondary_x[_qp_neighbour]
````

Which means I should somehow know the neighboring points of the current point and extract information within the same calculation. I know for calculating displacement jump we have `coupledNeighborValue`, but this may not be the case here since the information is stored in separate variables. I'll be appreciated for any workaround suggestions. Thanks in advance!

**Comment by hugary1995** [comment, marked as answer]
You should be able to couple neighbor reaction forces using `coupledNeighborValue` without any issue. You can use `_qp` to index both the current side and its neighbor side.

BTW, it's probably better to rename master/slave to primary/secondary here...

**Reply by chunhuizhao478** [reply]
Thanks for your reply. I think the following code does the things (correct me if I'm wrong), In the constructor:

````
_reaction_x(coupledValue("resid_primary_x")),
_reaction_neighbor_x(coupledNeighborValue("resid_secondary_x")),
_reaction_y(coupledValue("resid_primary_y")),
_reaction_neighbor_y(coupledNeighborValue("resid_secondary_y"))
````

In the `coupledNeighborValue`, I do the following the compute the difference

````
_reaction_x[_qp] - _reaction_neighbor_x[_qp]
````

And in the input file pass the corresponding auxvariable names. I once thought `coupledNeighborValue` should be acted on `resid_primary_x` in order to get its neighbor (`resid_secondary_x`) values, but that returns zero value, which is a bit confusing. Could you elaborate the logic here, or did I do something wrong?

**Reply by hugary1995** [reply]
This seems correct to me.
